client/client-lib/src/modules.rs

```diff
@@ -0,0 +1,260 @@
+use std::{collections::BTreeMap, io, sync::Arc};
+
+use impl_tools::autoimpl;
+
+use bitcoin::Amount;
+use thiserror::Error;
+
+/// Object-safe trait for things that can encode themselves
+///
+/// Like `rust-bitcoin`'s `consensus_encode`, but without generics,
+/// so can be used in `dyn` objects.
+pub trait Encodable {
+    fn consensus_encode(&self, writer: &mut dyn io::Write) -> Result<usize, std::io::Error>;
+}
+
+/// Something that can be an [`Input`] in a [`Transaction`]
+///
+/// General purpose code should use [`Input`] instead
+pub trait IInput: Encodable {}
+
+/// Something that can be an [`Output`] in a [`Transaction`]
+///
+/// General purpose code should use [`Output`] instead
+pub trait IOutput: Encodable {}
+
+/// An owned, immutable input to a [`Transaction`]
+#[autoimpl(Deref using self.0)]
+pub struct Input(Box<dyn IInput + Send + Sync + 'static>);
```

Comment:
This is a general pattern here. We introduce a `trait`, and then wrap it in a newtype. This way general purpose code treats everything like a value, and does not have to deal with `dyn` mess.

Every newtype `Deref`'s to `&dyn Trait` so the trait methods are easily accessible without too much boilerplate. We can use `autoimpl` to make this relatively painless.

More on `autoimpl` later (below).

Reply:
I went with `ISomething` and `Something` naming convention. Kind of stolen from Java. Not my fav. but in this case it really is this way - "an input" trait and "an input newtype wrapper over trait" are kind of the same thing. Just one is the interface, the other a wrapper. Since most code would work with a newtype, I want it to have the simple name. `SomethingTrait` feels meh, `SomethingIface`... maybe? Donno. Naming is hard.

Reply:
If the `ISomething` are things that the module writer needs to implement maybe we should prefix them with `Mod`?  Or don't prefix them to avoid confusing the module writer?

Reply:
If we don't prefix them then they will conflict with the concrete, but abstract `struct Something` that wraps then, which will be much more commonly used in all the code (so it should get a priority to get the nicest name).

`ModSomething` works with me, though might confuse people with Rust's `mod`. Maybe a longer `ModuleSomething`?